---
title: "RNAseq_052019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# Read in cell ranger output
read10x <- function(dir) {
  names <- c("barcodes.tsv", "features.tsv", "matrix.mtx")
   for (i in 1:length(names)) {
     R.utils::gunzip(paste0(dir, "/", names[i], ".gz"))
   }
  file.copy(paste0(dir, "/features.tsv"), paste0(dir, "/genes.tsv"))
  mat <- Seurat::Read10X(dir)
  file.remove(paste0(dir, "/genes.tsv"))
  for (i in 1:length(names)) {
    R.utils::gzip(paste0(dir, "/", names[i]))
  }
  mat
}
library(Seurat)
library(dplyr)
# setwd("/gpfs/analyses/maxsh/Allen/Isabel/190308_10x/Bulk")
```


```{r}
###### LOADING DATA #########
# Tell the program where the data directory is
data<-read10x("Data/Source")
```
Creating a seurat object which will be filled with your raw data
This seurat object is the core of the analysis
As a quality control measure, you need at least 3 cells and each cell needs to have 500 genes at bare minimum
```{r}
so <- CreateSeuratObject(counts = data, min.cells = 3, min.features = 500, 
                           project = "Isabel10xBulk")
# Looking to see how many genes are mitochondrial genes
mito.genes <- grep(pattern = "^mt-", x = rownames(x = GetAssayData(so)), value = TRUE)
percent.mito <- Matrix::colSums(GetAssayData(so, slot = "counts")[mito.genes, ])/Matrix::colSums(GetAssayData(so, slot = "counts"))
# Adding the mitochondrial information to the seurat object
so$percent.mito <- percent.mito
```


```{r}
######## BASIC QC ###########
# If not using as an R markdown document, use PDF function to generate PDF figures as output
pdf(file = "QC.pdf",width = 12,height = 12,pointsize=10)
VlnPlot(object = so, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), ncol = 3)
# If you use PDF function, include dev.off() so that the program knows you're done with the PDF
dev.off()
pdf(file = "QC2.pdf",width = 12,height = 12,pointsize=10)
# par(mfrow = c(1, 2))
FeatureScatter(object = so, feature1 = "nCount_RNA", feature2 = "percent.mito", pt.size = 0.5)
FeatureScatter(object = so, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", pt.size = 0.5)
dev.off()
```


```{r}
######## FILTERING ##########
# MANUALLY INPUT VALUES
# Filtering out lowly and too highly expressed genes, as well as cells where a high amount of expression comes from mitochondrial genes
so <- subset(so, subset = nFeature_RNA > 200 & nFeature_RNA < 5000 & percent.mito < .15)
```


```{r}
######## NORMALIZATION #####
so <- NormalizeData(so)
```


```{r}
######## VAR GENE IDENTIFICATION ####
# We care about genes that are highly variable between cells so we're focusing on those
pdf(file = "VarGenes.pdf", width = 12, height = 12, pointsize = 10)
so <- FindVariableFeatures(object = so, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(so), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(so)
plot2 <- LabelPoints(plot = plot1, points = top10)
# CombinePlots(plots = list(plot1, plot2))
plot1
plot2
dev.off()
```




```{r}
######## REMOVING UNWANTED VARIATION ###
# We're scaling and also regressing out the information we don't care about, in this case mitochondrial gene expression and the number of cells being different between groups
all.genes <- rownames(so)
so <- ScaleData(object = so, vars.to.regress = c("nCounts_RNA", "percent.mito"), features = all.genes)
```

```{r}
# This takes a long time, so if you'd like to save the R output for later use run this after the above cell has run
saveRDS(so, "Data/Seurat_Scaled.RDS")
```

```{r}
# When you come back to the data, run this
so <- readRDS("Data/Seurat_Scaled.RDS")
library(Seurat)
library(dplyr)
```


```{r}
######## PCA #######
# Running a PCA on the expression data to isolate the features that are most variable
# A way to visualize the data
so <- RunPCA(so, features = VariableFeatures(object = so))
print(so[["pca"]], dims = 1:5, nfeatures = 5)
```

```{r}
# Visualizing the most highly variable genes that are contributing to the principal components
VizDimLoadings(so, dims = 1:2, reduction = "pca")
```

```{r}
# Plotting the PCA
pdf(file = "PCA.pdf", width = 16, height = 12, pointsize = 36)
DimPlot(so, reduction = "pca")
dev.off()
```

Now we are going to plot the PCA data in three ways to identify the true dimensionality of the dataset. Jackstraw analysis is going to be very time-consuming, but definitely use the other two at least to be able to see what your data/PCA is looking like
```{r}
# Now we do a heat map of the first 5 PCs to take a peek at the most highly variable features
DimHeatmap(so, dims = 1:18, cells = 500, balanced = TRUE)
# Take a look at the genes on the right side of the graphs and see if you can spot your genes of interest, i.e. known astrocyte genes. They should be on the first few PCs, if not then there is something not so great about the data and it needs further optimization. 
```

```{r}
# Doing a jack straw analysis to randomly permute the data with resampling in order to generate a null distribution. 
# This takes a long time so we're commenting it out, but with more computing power we will do it. 
# so <- JackStraw(so, num.replicate = 100)
# so <- ScoreJackStraw(so, dims = 1:20)
# JackStrawPlot(so, dims = 1:15)
```

```{r}
# Making an elbow plot to see how many PCs to include. This is a heuristic, but you can often see by eye where it drops off.
pdf(file = "ElbowPlot.pdf", width = 16, height = 12, pointsize = 96)
ElbowPlot(so)
dev.off()
```


```{r}
# Again, if you'd like to save the R output for later, run this after the above cell has run
saveRDS(so, "Data/Seurat_PCA.RDS")
```

```{r}
# When you come back to the data, run this
so <- readRDS("Data/Seurat_PCA.RDS")
library(Seurat)
library(dplyr)
```

Now we can decide how many PCs to include in our analysis! Err on the higher side here. 

```{r}
########## Clustering ###########
# Change the "dims = " argument to whatever number of PCs you decide on
so <- FindNeighbors (so, dims = 1:15)
# The resolution value decides how "granular" your analysis is when you cluster. This should change depending on how many cells you have. 
so <- FindClusters (object = so, resolution = 0.1)
# Check out what cluster your first few cells are in if you want
# head(Idents(so), 5)
# Run a tSNE
so <- RunTSNE(object = so, dims = 1:15)
# so <- FindClusters(object = so, reduction.type = "pca", dims.use = 1:15, 
#                      resolution = 1.0, print.output = 0, save.SNN = TRUE)
# so <- RunTSNE(object = so, dims.use = 1:15, do.fast = TRUE)
```


```{r}
pdf(file = "TSNEPlot.pdf", width = 16, height = 12, pointsize = 24)
TSNEPlot(object = so)
dev.off()
```



```{r}
# Again, if you'd like to save the R output for later, run this after the above cell has run
saveRDS(so, "Data/Seurat_Clust.RDS")
```

```{r}
# When you come back to the data, run this
so <- readRDS("Data/Seurat_Clust.RDS")
library(Seurat)
library(dplyr)
```


```{r}
########## Coloring by average marker expression #######

## let's define the markers
markers=list()
markers$microglia=c("Tmem119","C1qa","C1qc","C1qb","Csf1r","Laptm5","Cd68","Olfml3","Tyrobp","Fcgr3","Alox5ap","Cx3cr1","P2ry13","Hexb","Lst1","Ctss","Coro1a","Ly86","Rtn4rl1","Rgs10","Rnase4","Kcnk12","Sla","Egr2","Ptgs1","Trpv2","Arhgdib","Gpr34","P2ry12","Rgs14","Ptk2b","Rassf5","Spp1")
markers$neuron=c("Reln","Slc17a6","Sst","5330417C22Rik","Dlx1","Npy","Nos1","Islr2","Dlx1as","Pnoc","Grem2","Clstn2","Npas4","Grm2","Nell1","Tmem130","Glra2","Slc32a1","Tbr1","Trim66","Lhx6","Bcl11a","Dpysl5","L1cam","Dync1i1","Tac2","Gpr83","Stmn2","Syt1","Crhbp","Tubb3","Rian","Kcnc2","Vgf","Bcl11b","Syt4","Cdh8","Kctd8","Gpr12","Elavl2","Synpr","Celsr3","Lrrc3b","Crh","Cpne5","Neurod1","Crmp1","Myt1l","Ina","Rgs8","Cpne4","Calb2","Gap43","Cacng2","Kcnip4","Eomes","Scg2","Adcyap1","Ablim3","Gpr22","Gabrg2","Rims3","Car10","Gpr26","Atp1a3","Rem2","Slc6a7","Gad1","Usp29","Cyp4x1","Nrn1","Kcnq5","Cbln4","Kcns2","Efna3","AW551984","Gabra5","Baiap3","Cacna1b","Zcchc12","Zic1","Sv2b","Hspa12a","Rhov","Gria1","Kcna4","Cntnap4","Hcn1","Trhde","Cacna2d1","Doc2b","A830018L16Rik")
markers$opc=c("Lhfpl3","3110035E14Rik","Dcn","Lrrtm3","Sulf1","3632451O06Rik","Gria3","Prkg2","Tmem132d","C1ql3","Nxph1","Gpr17","Cntn4","Cobl","Hrasls","Slitrk3")
markers$oligo=c("Mog","Cldn11","Mobp","Mbp","Plp1","Mag","Mal","Ppp1r14a","Fa2h","Gsn","Gpr62","Tnni1","Kndc1","Ndrg1","Ugt8a","Tspan2","Adamts4","1700047M11Rik","Adssl1","Tmem117","Tmeff2","Sh3gl3","Plekhh1","Kcna1","Bcas1","Cpm")
markers$endo=c("Car4","Itm2a","Ly6a","Egfl7","Cd34","Kitl","Sgpp2","Cdc42ep3","Ebf1","Ankrd37","Arhgef15","Akap2","Myo1b","Mgp","Gcnt2","Tiam1","Plk2","Slc39a10","Smoc2","Abcg2","Grb10","Pde5a","Bok","Myl4","Rasgrp3","Vtn","Tmsb10","9430020K01Rik","Anxa3","Rassf3")
markers$astrocyte=c("Aqp4","Bmpr1b","Plcd4","Ppp1r3c","Slc30a10","Itih3","Slc14a1","Atp13a4","Grm3","Paqr6","Fgfr3","Slc4a4","Itga7","Aldh1l1","St6galnac5","Cbs","Phkg1","Slc6a11","Slc25a34","Sorcs2","Cyp4f15","Entpd2","Grhl1","Dio2","Sox9","Slc15a2","Slc7a2","AI464131","Slc7a10","Slc39a12","Mlc1","Ppp1r3g","Ephx2","Slc1a3","2900052N01Rik","Vcam1","Adhfe1","Rorb","Daam2","Rgs20","Abcd2","Lrig1","Slc7a11","Cldn10","Elovl2","Igsf1","Slc25a18","F3","Fabp7","Emp2","Pdk4","Acsbg1","Kcne1l","Igfbp2","Gpc4","Hes5","Slc1a2","Tlcd1","Dzip1l","Gfap","Tst","Rfx4","Grin2c","Slc13a3","Fjx1","Megf10","Slc27a1","Pla2g7","Mfge8","Pax6","Rdh5","Cyp4f14","Hsd11b1","Lcat","Gpam","Agt","Olfml1","Dbx2","Hapln1","Amot","Gjb6","Mtmr11","Ccnd2","Mapk4","Acsl6","Sfxn5","Ptch1","Htra1","Rasa2","Gabrg1","Id4","Scara3","Atp1b2","Acot11","Klf15","Tpbg","Ptprz1","Fgfr1","Vegfa","Fxyd1","Lonrf3","Nhsl1")
# so = AddMetaData (so, Matrix::colMeans(so@scale.data[rownames(so@scale.data) %in% markers$microglia, ]),col.name = "Avg. Microglia")
# AddMetaData(object = object, metadata = vector, col.name = "name") // object$name <- vector
# object@scale.data // GetAssayData(object = object, slot = "scale.data")
scaledso <- GetAssayData ( object = so, slot = "scale.data")
so$Avg.Microglia <- (Matrix::colMeans(scaledso [rownames(so) %in% markers$microglia, ]))
so$Avg.Neuron <- (Matrix::colMeans(scaledso [rownames(so) %in% markers$neuron, ]))
so$Avg.Astrocyte <- (Matrix::colMeans(scaledso [rownames(so) %in% markers$astrocyte, ]))
so$Avg.OPC <- (Matrix::colMeans(scaledso [rownames(so) %in% markers$opc, ]))
so$Avg.Oligo <- (Matrix::colMeans(scaledso [rownames(so) %in% markers$oligo, ]))
so$Avg.Endo <- (Matrix::colMeans(scaledso [rownames(so) %in% markers$endo, ]))
# so=AddMetaData(so,Matrix::colMeans(so@scale.data[rownames(so@scale.data) %in% markers$neuron, ]),col.name = "Avg. Neuron")
# so=AddMetaData(so,Matrix::colMeans(so@scale.data[rownames(so@scale.data) %in% markers$opc, ]),col.name = "Avg. OPC")
# so=AddMetaData(so,Matrix::colMeans(so@scale.data[rownames(so@scale.data) %in% markers$oligo, ]),col.name = "Avg. Oligo")
# so=AddMetaData(so,Matrix::colMeans(so@scale.data[rownames(so@scale.data) %in% markers$endo, ]),col.name = "Avg. Endo")
# so = AddMetaData(so,Matrix::colMeans(so@scale.data[rownames(so@scale.data) %in% markers$astrocyte, ]),col.name = "Avg. Astrocyte")
```

```{r}
# Let's make TSNEs of our data with clusters labelled that we think should be present
FeaturePlot(object = so, features = c("Avg.Microglia","Avg.Neuron", "Avg.OPC", "Avg.Oligo", "Avg.Endo", "Avg.Astrocyte"), cols = c("lightgrey","red"), reduction = "tsne")
```


```{r}
# Let's make a TSNE with just astrocyte markers labeled
FeaturePlot(object = so, features = markers$astrocyte, cols = c("lightgrey","red"), reduction = "tsne")
```


```{r}
######## Saving the object #####
saveRDS(so, file = "Data/Seurat_Final.RDS")
```
```{r}
# When you come back to the data, run this
so <- readRDS("Data/Seurat_Final.RDS")
library(Seurat)
library(dplyr)
```


```{r}
######## FINDING MARKERS #######
so.markers <- FindAllMarkers(object = so, only.pos = TRUE, min.pct = 0.1)
# Top 10 most upregulated genes per cluster
top10 <- so.markers %>% group_by(cluster) %>% top_n(10, avg_logFC)
```

```{r}
DoHeatmap(object = so, features = top10$gene)
```


```{r}
so2 <- WhichCells(object = so, idents = c(0), invert = TRUE)
```

```{r}
pdf("Top10Genes.pdf", width = 16, height = 12)
DoHeatmap(object = so[, so2], features = top10$gene[-seq(1, 10)])
dev.off()
```


```{r}
clust <- table(so@active.ident)
clust
```

```{r}
require(scales)

# Create vector with levels of object@ident
identities <- levels(so@active.ident)

# Create vector of default ggplot2 colors
my_color_palette <- hue_pal()(length(identities))
pdf(file = "BarPlotBulk.pdf", width = 16, height = 12, pointsize = 36)
barplot(clust, col = my_color_palette, xlab = "Cluster", ylab = "Number of cells")
dev.off()
```

